### ISSUE TITLE
Support new architecture

### ISSUE THREAD

**Issue by nandorojo** [dominicstop/react-native-ios-context-menu#100]
Lifting #84 into a new issue.

React Native 0.74 is coming with bridgeless mode and improvements for the new architecture. It appears that Expo is finally adopting it and they have a plan to test libraries for it.

Here brent explains how to test New Architecture support with an Expo Module (it's easy): https://x.com/notbrent/status/1774931733194088465?s=20

TLDR: install `expo-build-properties` into the app you're testing, and add it to your app.json, setting `newArchEnabled: true`.

```json
{
  "expo": {
    "name": "try-zeego",
    "slug": "try-zeego",
    "icon": "./assets/icon.png",
    "splash": {
      "image": "./assets/splash.png",
      "resizeMode": "contain",
      "backgroundColor": "#ffffff"
    },
    "ios": {
      "supportsTablet": true,
      "bundleIdentifier": "com.example.with-new-arch"
    },
    "android": {
      "package": "com.example.withnewarch"
    },
    "plugins": [
      [
        "expo-build-properties",
        {
          "ios": {
            "newArchEnabled": true
          },
          "android": {
            "newArchEnabled": true
          }
        }
      ]
    ]
  }
}
```

[Here is a reproduction](https://github.com/brentvatne/with-zeego-new-arch) showing that this library does not currently work with the new architecture.

Basically the menu doesn't open.

# Solution

Fabric deprecates `findNodeHandle()`, and it looks like `react-native-ios-context-menu` [uses that.](https://github.com/search?q=repo%3Adominicstop%2Freact-native-ios-context-menu%20findnodehandle&type=code)

I think the solution would be to remove `findNodeHandle` in favor of the component's ref.

**Comment by fobos531** [comment]
> This also removes the bridge, so I think self.bridge would not be an option to use either.

About this, it seems that bridgeless mode still exposes a "bridge" in some way, probably to ensure maximum backwards compatibility

![CleanShot 2024-04-04 at 10 39 51@2x](https://github.com/dominicstop/react-native-ios-context-menu/assets/19533289/0759c3a4-a83b-4f10-9f8e-2b41bf2a4339)

So hopefully it might not be a big hindrance

**Comment by nandorojo** [comment]
Turns out `self.bridge` isn't in the source anyway so I think it's a moot point. We will have figure out the findNodeHandle though

**Comment by fobos531** [comment]
> Turns out `self.bridge` isn't in the source anyway so I think it's a moot point. We will have figure out findNodeHandle though

`self.bridge` is not directly utilized by `react-native-ios-context-menu`, but it looks like it is heavily utilized by `react-native-ios-utilities`, which `react-native-ios-context-menu`. Either way, I agree, both these concerns should be addressed to ensure 0.74 + bridgeless compatibility.

cc @dominicstop

**Comment by nandorojo** [comment]
Oh, got it, good to know. Well at least we know that will work then.

**Comment by dominicstop** [comment]
# 


Hi, thanks for tagging me, the issue got buried in my inbox, so I apologize for the very late reply
<br>
It looks like I’ll need to do some refactors + run some experiments to support the new architecture — the one thing i’m worried about the most is that the workarounds/hacks i’ve used in the past won’t work anymore w/ fabric...
<br>
I’m currently working on [making a component that wraps `UITableView`](https://github.com/dominicstop/react-native-ios-list-view/blob/master/example/src/examples/TableViewTest02/TableViewTest02.tsx), but ran into some trouble due to the delay/async communication between native and JS.
<br>
Below is a quick demo:

https://github.com/dominicstop/react-native-ios-context-menu/assets/18517029/c9d76b04-581a-4991-ba74-a9bddde4afe9

The demo shows the ff. things:
* Using regular RN components for the table view’s cell content.
* drag and drop, and cell re-ordering (via `UITableViewDiffableDataSource` + `UIDragDelegate`).
* Cell reuse (i.e. the state of the RN cell content is preserved across rows since it’s being reused, hence the counter stays the same across rows, but the content changes).
* Dynamic cell height (i.e self-sizing cells).
* Mapping JS data to the table view data source.
* etc..
<br>
i'm not sure yet, but i think the new architecture + expo API's will fix this...

**Comment by fobos531** [comment]
> As for the bridge usage, i'm not sure yet if there's a corresponding replacement for all the API's i'm using...

@dominicstop I haven't check in full detail, but it seems like you're not using any APIs that the RN team has explicitly specified that the RCTBridgeProxy does not support. So, it may work "out of the box" as long as you switch to using this RCTBridgeProxy instead of the RCTBridge

**Comment by fobos531** [comment]
Hello @dominicstop 

As a heads up, RN 0.74 was officially released yesterday, which means it is likely that there will be an increase in the number of users actively experimenting with the new architecture, which will probably bring more attention to this issue.

**Comment by dominicstop** [comment]
@fobos531 hello, thank you for the gentle warning/reminder; i made 2 new branches so i can start working on this (i.e. [`react-native-ios-utilities/wip-new-arch`](https://github.com/dominicstop/react-native-ios-utilities/tree/wip-new-arch), and [`react-native-ios-context-menu/wip-new-arch`](https://github.com/dominicstop/react-native-ios-context-menu/tree/wip-new-arch)).
<br>
Unfortunately, i won't be able to support the "aux. previews" for now (i'm hoping to fix support for it in future releases); but i'll be able to make sure all the other stuff work (my best estimate is around 2-3 days).
<br>
i understand your frustration, and i apologize. In summary, i'm currently in between places (and i lost my current job that allows me to work on OSS full-time). For the foreseeable future, i'm going to be supporting this library in my free time only (as a hobby), since i'm going to go back to my non-tech job again. But for now, i still have a couple of days left to work on supporting the new architecture.
<br>
**TLDR**: I'm working on it, and i'm hoping to finish it before the end of the month (i'll update this thread once i'm able to make a pre-release version for testing).
<br>
thank you for understanding ✨

**Comment by 1kuko3** [comment]
An SDK 58 / RN 0.88 RC update for this architecture discussion: I isolated a compile blocker in `react-native-ios-utilities` 5.2.0, which Context Menu depends on, and reported it with a standalone reproduction:

https://github.com/dominicstop/react-native-ios-utilities/issues/45

A clean Expo 58.0.0-preview.4 / RN 0.88.0-rc.1 iOS build cannot resolve `RCTImageLoaderWithAttributionProtocol` in Swift. This is additional to the older `RCTRootContentView` problem tracked in the Utilities repository.

There is also a small source change to consider in Context Menu 3.2.1: both `RNIContextMenuViewManager.mm` and `RNIContextMenuButtonViewManager.mm` still have an unconditional `#import "RCTBridge.h"`, although their bridge-using `view` methods are already inside `#ifndef RCT_NEW_ARCH_ENABLED`. Our local patch moves the include under the same guard and uses the qualified path:

```objc
#ifndef RCT_NEW_ARCH_ENABLED
#import <React/RCTBridge.h>
#endif
```

The same unqualified-header problem has been reported independently for Menu on plain RN 0.87: https://github.com/react-native-menu/menu/issues/1221.

The fresh minimal build above verifies the Utilities failure only. I haven't completed a separate Context Menu build or menu-interaction test with these RC versions, so the import change is a source-level finding, not a claim that it completes RN 0.88 support. I'm keeping the older subtitle availability issue out of this update.
